_posts/2020-04-27-ensuring-resolvers-aren't-rejected.md

```diff
@@ -0,0 +1,77 @@
+
+![Console - Edit Resolver ]({{ site.url }}/img/editResolver-console.png)
+
+Having found myself recently in this exact predicament, I embarked on a search for "a better way". I came across some packages deep within the murky depths of the Amplify framework and discovered that by reverse engineering one or two of these packages I could reproduce the VTL environment in development. This allowed me to develop VTL templates test first and get a super fast feedback loop into the JSON documents my templates were outputting before I deployed them anywhere.
+
+Below is a trivial example where I use a VTL template to create a JSON document that will perform a DynamoDB PutItem operation. There is no syntax highlighting support for VTL in markdown so I have included an image instead.
+
+![vtl template]({{ site.url }}/img/createThingVtlTemplate.png)
+
+The test will look something like the following.
+
+```typescript
+import * as utils from 'amplify-appsync-simulator/lib/velocity/util'
+import { map } from 'amplify-appsync-simulator/lib/velocity/value-mapper/mapper'
+import { Compile, parse } from 'amplify-velocity-template'
+import * as fs from 'fs'
+import * as path from 'path'
+import { expect } from 'chai'
+
+describe('Create Thing', () => {
+  it('should create the correct id value and the DynamoDB data type descriptor object ', () => {
+    const templatePath = path.resolve(__dirname, '../templates/createSMP-request.vtl')
+    const createSMPTemplate = fs.readFileSync(templatePath, { encoding: 'utf8' })
+    const ast = parse(createSMPTemplate)
+    const compiler = new Compile(ast, {
+      valueMapper: map,
+      escape: false
+    })
+    const args = {
+      thing: {
+        attribute1: '000',
+        attribute2: '123',
+        attribute3: '456'
+      }
+    }
+    const context = createVtlContext(args)
+    const result = JSON.parse(compiler.render(context))
+    expect(result.key.id.S.startsWith(`thing_${args.thing.attribute1}`)).to.be.true
+    const uuidRegex = /[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}/
+    expect(result.key.id.S).to.match(uuidRegex)
+  })
+})
+
+function createVtlContext<T> (args: T) {
+  const util = utils.create([], new Date(Date.now()), Object())
+  const context = {
+    args,
+    arguments: args
+  }
+  return {
+    util,
+    utils: util,
+    ctx: context,
+    context
+  }
+}
+```
+
+Having the ability to unit test my resolver mapping templates means I can use these tests to aid my development efforts as well as run these tests as part of a test suite in my CI/CD pipeline, making sure that these templates at least make sense before deploying new resolvers to my GraphQL API.
```

Comment:
is there more?
Needs a conclusion of some kind and a hook.. so they have read it, what next?